src/utils/EventUtils.cpp

```diff
@@ -0,0 +1,565 @@
+  auto data = std::get<0>(evData.emplace(std::piecewise_construct,
+                                         std::forward_as_tuple(event->name),
+                                         std::forward_as_tuple(event->name)));
+  data->second.put(event);
+}
+
+
+void RankData::addEventData(EventData ed)
+{
+  evData.emplace(ed.getName(), std::move(ed));
+}
+
+
+void RankData::normalizeTo(sys_clk::time_point t0)
+{
+  auto const delta = initializedAt - t0; // duration that this rank initialized after the first rank
+  assertion(t0 <= initializedAt); // t0 should always be before or equal my init time
+
+  for (auto & events : evData) {
+    for (auto & sc : events.second.stateChanges) {
+      auto & tp = sc.second;
+      tp = stdy_clk::time_point(tp - initializedAtTicks + delta);
+      // Deactivated, as it fails for the precice tests, likely because Events are used before initialize
+      // assert(tp.time_since_epoch().count() >= 0); // Trying to do normalize twice?
+    }
+  }
+}
+
+void RankData::clear()
+{
+  evData.clear();
+}
+
+stdy_clk::duration RankData::getDuration() const
+{
+  if (isFinalized)
+    return finalizedAtTicks - initializedAtTicks;
+  else
+    return stdy_clk::now() - initializedAtTicks;
+}
+
+
+
+// -----------------------------------------------------------------------
+
+
+EventRegistry & EventRegistry::instance()
+{
+  static EventRegistry instance;
+  return instance;
+}
+
+void EventRegistry::initialize(std::string applicationName, std::string runName, MPI_Comm comm)
+{
+  this->applicationName = applicationName;
+  this->runName = runName;
+  this->comm = comm;
+
+  localRankData.initialize();
+
+  globalEvent.start(false);
+  initialized = true;
+}
+
+void EventRegistry::finalize()
+{
+  globalEvent.stop();
+  localRankData.finalize();
+
+  for (auto & e : storedEvents)
+    e.second.stop();
+
+  if (initialized) // this makes only sense when it was properly initialized
+    normalize();
+
+  collect();
+
+  initialized = false;
+}
+
+void EventRegistry::clear()
+{
+  localRankData.clear();
+  globalRankData.clear();
+  storedEvents.clear();
+}
+
+void EventRegistry::signal_handler(int signal)
+{
+  if (initialized) {
+    finalize();
+    printAll();
+  }
+}
+
+void EventRegistry::put(Event* event)
+{
+  localRankData.put(event);
+}
+
+Event & EventRegistry::getStoredEvent(std::string const & name)
+{
+  // Reset the prefix for creation of a stored event. Using prefixes with stored events is possible
+  // but leads to unexpected results, such as not getting the event you want, because someone else up the
+  // stack set a prefix.
+  auto previousPrefix = prefix;
+  prefix = "";
+  auto insertion = storedEvents.emplace(std::piecewise_construct,
+                                        std::forward_as_tuple(name),
+                                        std::forward_as_tuple(name, false, false));
+
+  prefix = previousPrefix;
+  return std::get<0>(insertion)->second;
+}
+
+
+void EventRegistry::printAll()
```

Comment:
This function looks weird to me as it says _print_, but it does _print and writes the log file only from master_ .
Can you please split it up?

Here a suggestion:
```cpp
void EventRegistry::printSummary() { 
  writeSummary(std::cout);
}

void EventRegistry::writeLog() { 
  std::string logFile;
  if (applicationName.empty())
    logFile = "Events.json";
  else
    logFile = applicationName + "-events.json";

  std::ofstream ofs(logFile);
  writeJSON(ofs);
 }

void EventRegistry::logFromMaster() {
  int myRank;
  MPI_Comm_rank(comm, &myRank);

  if (myRank == 0) {
    printSummary();
    writeLog();
  }
}
```

Reply:
I am not to eager to split it up, since I want /one/ function that is called after finalize and does all the log writing / printing.
`writeTimings` and `writeLog` are ok, they both write to their argument `ostream out`.

The equivalance of master and rank 0 is a precice only thing, and actually, it shouldn't be here either.

Reply:
_"Split up"_ was a poor formulation.
I really meant that it would be a good idea to _refactor_ those independent parts out of that one function. This makes the code more generic and easier to read as the responsibilities of each function are clearer.

There are 5 concerns here:
1) __generate a json representation__ of the recorded events and write them to a generic ostream  
`EventRegistry::writeJSON(osteam)` (suggested name change in review)
2) __generate a summary__ of the timings of the recorded events and write them to a generic ostream  
Calls `EventRegistry::writeSummary(ostream)`
3) generate and __write the log file__   (suggested name change in review)
`void EventRegistry::writeLog()` (should probably be `writeLogFile()`)  
Calls `EventRegistry::writeJSON(ofsteam)`
4) __print the summary__ to std::out  
`void EventRegistry::printSummary()`
Calls `EventRegistry::writeSummary(std::cout)`
5) Print summary and write log file from master (__preCICE specific__)  
`void logFromMaster()`

_Note: the generation of the file name is a 6th concern, but not important here._


> I want /one/ function that is called after finalize and does all the log writing / printing.

That would be a call to `EventRegistry::instance()::logFromMaster()`.


Now there is only one minor precice specific part of the EventTimings that could be eventually refactored out to further decouple the project from precice:
```cpp
namespace precice::utils {
void logEventsFromMaster() {
  if(utils::Parallel::getProcessRank() != 0) {
    return;
  }
  auto& instance = EventRegistry::instance();
  instance.printSummary();
  instance.writeLog();
}
}
```